Source/WebCore/workers/service/background-fetch/BackgroundFetch.cpp

```diff
@@ -269,14 +315,61 @@ void BackgroundFetch::Record::didSendData(uint64_t size)
         m_fetch->didSendData(size);
 }
 
+// https://wicg.github.io/background-fetch/#extract-content-range-values
+static ParsedContentRange extractContentRangeValues(const ResourceResponse& response)
+{
+    return ParsedContentRange(response.httpHeaderField(HTTPHeaderName::ContentRange));
+}
+
+// https://wicg.github.io/background-fetch/#validate-a-partial-response
+static bool validatePartialResponse(size_t rangeStart, const ResourceResponse& partialResponse, const ResourceResponse& previousResponse)
+{
+    auto parsedContentRange = extractContentRangeValues(partialResponse);
+    if (!parsedContentRange.isValid())
+        return false;
+
+    if (static_cast<size_t>(parsedContentRange.firstBytePosition()) != rangeStart)
+        return false;
+
+    auto previousETag = previousResponse.httpHeaderField(HTTPHeaderName::ETag);
+    if (!previousETag.isEmpty() && previousETag != partialResponse.httpHeaderField(HTTPHeaderName::ETag))
+        return false;
+
+    auto previousLastMotified = previousResponse.httpHeaderField(HTTPHeaderName::LastModified);
+    if (!previousLastMotified.isEmpty() && previousLastMotified != partialResponse.httpHeaderField(HTTPHeaderName::LastModified))
+        return false;
+
+    if (previousResponse.httpStatusCode() == 206) {
+        auto parsedPreviousContentRange = extractContentRangeValues(previousResponse);
+        if (!parsedPreviousContentRange.isValid())
+            return false;
+        if (parsedPreviousContentRange.instanceLength() != ParsedContentRange::unknownLength && parsedPreviousContentRange.instanceLength() != parsedContentRange.instanceLength())
+            return false;
+    }
+    return true;
+}
+
 void BackgroundFetch::Record::didReceiveResponse(ResourceResponse&& response)
 {
+    bool shouldClearResponseBody = false;
+    if (response.httpStatusCode() == 206) {
+        if (!validatePartialResponse(m_responseDataSize, response, m_response)) {
+            didFinish(ResourceError { String { }, 0, response.url(), "Validation of partial response failed"_s, ResourceError::Type::AccessControl });
+            return;
```

Comment:
Do we need to clear the partial response on disk if validation fails?

Reply:
Good point, we do not need to, but it would reduce disk usage.
Spec allows it I think since, after this error, the record is marked as 'fetch-error' and we no longer can expose the response nor its body.
I added a comment in https://bugs.webkit.org/show_bug.cgi?id=253813 to implement this